Description: Don't kill the container just bc console got hangup
Author: "Daniel P. Berrange" <berrange@redhat.com>
Forwarded: no
Bug-Ubuntu: https://bugs.launchpad.net/bugs/668369

Index: libvirt-0.8.8-0ubuntu1/src/lxc/lxc_controller.c
===================================================================
--- libvirt-0.8.8-0ubuntu1.orig/src/lxc/lxc_controller.c	2011-02-08 19:29:41.000000000 -0600
+++ libvirt-0.8.8-0ubuntu1/src/lxc/lxc_controller.c	2011-02-23 08:51:56.751332581 -0600
@@ -323,6 +323,18 @@
           || errnum == EWOULDBLOCK);
 }
 
+static bool
+lxcPidGone(pid_t container)
+{
+    waitpid(container, NULL, WNOHANG);
+
+    if (kill(container, 0) < 0 &&
+        errno == ESRCH)
+        return true;
+
+    return false;
+}
+
 /**
  * lxcControllerMain
  * @monitor: server socket fd to accept client requests
@@ -340,7 +352,8 @@
 static int lxcControllerMain(int monitor,
                              int client,
                              int appPty,
-                             int contPty)
+                             int contPty,
+			     pid_t container)
 {
     int rc = -1;
     int epollFd;
@@ -446,7 +459,13 @@
                         ++numActive;
                     }
                 } else if (epollEvent.events & EPOLLHUP) {
-                    DEBUG("EPOLLHUP from fd %d", epollEvent.data.fd);
+                    if (lxcPidGone(container))
+                        goto cleanup;
+                    curFdOff = epollEvent.data.fd == appPty ? 0 : 1;
+                    if (fdArray[curFdOff].active) {
+                        fdArray[curFdOff].active = 0;
+                        --numActive;
+                    }
                     continue;
                 } else {
                     lxcError(VIR_ERR_INTERNAL_ERROR,
@@ -485,7 +504,9 @@
                 --numActive;
                 fdArray[curFdOff].active = 0;
             } else if (-1 == rc) {
-                goto cleanup;
+		if (lxcPidGone(container))
+                    goto cleanup;
+                continue;
             }
 
         }
@@ -683,7 +704,7 @@
     if (lxcControllerClearCapabilities() < 0)
         goto cleanup;
 
-    rc = lxcControllerMain(monitor, client, appPty, containerPty);
+    rc = lxcControllerMain(monitor, client, appPty, containerPty, container);
 
 cleanup:
     VIR_FREE(devptmx);
